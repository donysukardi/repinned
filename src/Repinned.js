import { Component } from 'react'
import PropTypes from 'prop-types'
import shallowequal from 'shallowequal'
import 'raf/polyfill'

import shouldUpdate from './shouldUpdate'
import { noop, unwrapArray } from './utils'

class Repinned extends Component {
  constructor(props) {
    super(props)
    // Class variables.
    this.currentScrollY = 0
    this.lastKnownScrollY = 0
    this.scrollTicking = false
    this.resizeTicking = false
    this.state = {
      state: props.forcePin ? 'pinned' : 'unfixed',
      shouldAnimate: false,
      height: 0,
    }
  }

  setRef = ref => {
    this.inner = ref
  }

  setHeightOffset = () => {
    this.setState(state =>
      state.height === this.inner.offsetHeight
        ? null
        : { height: this.inner.offsetHeight }
    )

    this.resizeTicking = false
  }

  getScrollY = () => {
    const parent = this.props.parent()
    /* istanbul ignore else  */
    if (parent.pageYOffset !== undefined) {
      return parent.pageYOffset
    } else if (parent.scrollTop !== undefined) {
      return parent.scrollTop
    }
    /* istanbul ignore next line */
    return (
      document.documentElement ||
      document.body.parentNode ||
      document.body
    ).scrollTop
  }

  getViewportHeight = () =>
    /* istanbul ignore next line */
    window.innerHeight ||
    document.documentElement.clientHeight ||
    document.body.clientHeight

  getDocumentHeight = /* istanbul ignore next */ () => {
    const { body, documentElement } = document
    return Math.max(
      body.scrollHeight,
      documentElement.scrollHeight,
      body.offsetHeight,
      documentElement.offsetHeight,
      body.clientHeight,
      documentElement.clientHeight,
    )
  }

  getElementPhysicalHeight = elm => Math.max(elm.offsetHeight, elm.clientHeight)

  getElementHeight = elm =>
    Math.max(elm.scrollHeight, elm.offsetHeight, elm.clientHeight)

  getScrollerPhysicalHeight = () => {
    const parent = this.props.parent()

    return parent === window || parent === document.body
      ? /* istanbul ignore next */ this.getViewportHeight()
      : this.getElementPhysicalHeight(parent)
  }

  getScrollerHeight = () => {
    const parent = this.props.parent()

    return parent === window || parent === document.body
      ? /* istanbul ignore next */ this.getDocumentHeight()
      : this.getElementHeight(parent)
  }

  isOutOfBound = currentScrollY => {
    const pastTop = currentScrollY < 0

    const scrollerPhysicalHeight = this.getScrollerPhysicalHeight()
    const scrollerHeight = this.getScrollerHeight()

    const pastBottom = currentScrollY + scrollerPhysicalHeight > scrollerHeight

    return pastTop || pastBottom
  }

  handleScroll = () => {
    /* istanbul ignore else  */
    if (!this.scrollTicking) {
      this.scrollTicking = true
      window.requestAnimationFrame(this.update)
    }
  }

  handleResize = /* istanbul ignore next */ () => {
    /* istanbul ignore else  */
    if (!this.resizeTicking) {
      this.resizeTicking = true
      window.requestAnimationFrame(this.setHeightOffset)
    }
  }

  setNewState = state => {
    this.setState(prevState => ({
      state,
      shouldAnimate: prevState.state !== 'unfixed',
    }))
  }

  unpin = () => {
    this.props.onUnpin()
    this.setNewState('unpinned')
  }

  pin = () => {
    this.props.onPin()
    this.setNewState('pinned')
  }

  unfix = () => {
    this.props.onUnfix()
    this.setNewState('unfixed')
  }

  update = () => {
    this.currentScrollY = this.getScrollY()
    /* istanbul ignore else  */
    if (!this.isOutOfBound(this.currentScrollY)) {
      const { action } = shouldUpdate(
        this.lastKnownScrollY,
        this.currentScrollY,
        this.props,
        this.state,
        this.state.height
      )

      /* istanbul ignore else  */
      if (action === 'pin') {
        this.pin()
      } else if (action === 'unpin') {
        this.unpin()
      } else if (action === 'unfix') {
        this.unfix()
      }
    }

    this.lastKnownScrollY = this.currentScrollY
    this.scrollTicking = false
  }

  componentDidMount() {
    const { disabled, forcePin, parent: parentFn, calcHeightOnResize } = this.props
    this.setHeightOffset()

    const parent = parentFn()
    if (!disabled) {
      if(!forcePin) {
        parent.addEventListener('scroll', this.handleScroll)
      }
      if (calcHeightOnResize) {
        parent.addEventListener('resize', this.handleResize)
      }
    }
  }

  shouldComponentUpdate(nextProps, nextState) {
    return (
      !shallowequal(this.props, nextProps) ||
      !shallowequal(this.state, nextState)
    )
  }

  componentDidUpdate(prevProps) {
    const parent = this.props.parent()
    if(this.props.forcePin && !prevProps.forcePin) {
      this.pin()
      parent.removeEventListener('scroll', this.handleScroll)
    } else if(!this.props.forcePin && prevProps.forcePin) {
      parent.addEventListener('scroll', this.handleScroll)
    }

    if (this.props.disabled && !prevProps.disabled) {
      this.unfix()
      parent.removeEventListener('scroll', this.handleScroll)
      parent.removeEventListener('resize', this.handleResize)
    } else if (!this.props.disabled && prevProps.disabled) {
      parent.addEventListener('scroll', this.handleScroll)
      /* istanbul ignore else  */
      if (this.props.calcHeightOnResize) {
        parent.addEventListener('resize', this.handleResize)
      }
    }
  }

  componentWillUnmount() {
    const parent = this.props.parent()
    parent.removeEventListener('scroll', this.handleScroll)
    /* istanbul ignore else  */
    if (parent !== window) {
      window.removeEventListener('scroll', this.handleScroll)
    }
    parent.removeEventListener('resize', this.handleResize)
  }

  render() {
    const children = unwrapArray(this.props.children, noop)

    return children({
      setRef: this.setRef,
      ...this.state,
    })
  }
}

Repinned.propTypes = {
  parent: PropTypes.func,
  children: PropTypes.func.isRequired,
  disabled: PropTypes.bool,
  upTolerance: PropTypes.number,
  downTolerance: PropTypes.number,
  onPin: PropTypes.func,
  onUnpin: PropTypes.func,
  onUnfix: PropTypes.func,
  pinStart: PropTypes.number,
  forcePin: PropTypes.bool,
  calcHeightOnResize: PropTypes.bool,
}

Repinned.defaultProps = {
  parent: () => window,
  forcePin: false,
  disabled: false,
  upTolerance: 5,
  downTolerance: 5,
  onPin: noop,
  onUnpin: noop,
  onUnfix: noop,
  pinStart: 0,
  calcHeightOnResize: true,
}

export default Repinned
