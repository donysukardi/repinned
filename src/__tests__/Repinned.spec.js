import React from 'react'
import PropTypes from 'prop-types'
import { cleanup, render } from 'react-testing-library'
import 'jest-dom/extend-expect'

import Repinned from '../Repinned'

beforeEach(() => {
  jest
    .spyOn(global.window, 'requestAnimationFrame')
    .mockImplementation(cb => cb())
})

afterEach(() => {
  cleanup()
  global.window.requestAnimationFrame.mockRestore()
})

const getParent = () => ({
  addEventListener: jest.fn(),
  removeEventListener: jest.fn(),
  pageYOffset: 0,
  scrollHeight: 1200,
  offsetHeight: 50,
})

const RepinnedExample = ({ repinnedRef, ...props }) => (
  <div style={{ minHeight: '100vh' }}>
    <Repinned {...props} ref={repinnedRef}>
      {({ setRef, height, state }) => (
        <div
          data-testid="repinnedWrapper"
          style={{
            height,
          }}
        >
          <header ref={setRef} data-testid="headerContainer">
            <h1>This is header</h1>
            <div data-testid="RepinnedState">{state}</div>
          </header>
        </div>
      )}
    </Repinned>
    <div>This is some content</div>
  </div>
)

RepinnedExample.propTypes = {
  repinnedRef: PropTypes.func,
}

test('Repinned component renders properly', () => {
  const { getByTestId, container } = render(<RepinnedExample />)
  expect(getByTestId('RepinnedState')).toHaveTextContent('unfixed')
  expect(container.firstChild).toMatchSnapshot()
})

test('Repinned does not register on resize event listener if calcHeightOnResize is false', () => {
  const parent = getParent()
  render(<RepinnedExample parent={() => parent} calcHeightOnResize={false} />)

  expect(parent.addEventListener.mock.calls[0][0]).toBe('scroll')
  expect(parent.addEventListener.mock.calls.length).toBe(1)
})

test('Repinned didMount registers scroll event listener', () => {
  const parent = getParent()
  const { unmount } = render(<RepinnedExample parent={() => parent} />)

  expect(parent.addEventListener.mock.calls[0][0]).toBe('scroll')
  expect(parent.addEventListener.mock.calls[1][0]).toBe('resize')

  unmount()
  expect(parent.removeEventListener.mock.calls[0][0]).toBe('scroll')
  expect(parent.removeEventListener.mock.calls[1][0]).toBe('resize')
})

test('Repinned toggling forcePin on should remove event listeners', () => {
  const parent = getParent()
  const parentFn = () => parent

  const { rerender } = render(<RepinnedExample parent={parentFn} />)

  expect(parent.addEventListener.mock.calls[0][0]).toBe('scroll')

  rerender(<RepinnedExample parent={parentFn} forcePin />)

  expect(parent.removeEventListener.mock.calls[0][0]).toBe('scroll')
})

test('Repinned toggling forcePin off should add event listeners', () => {
  const parent = getParent()
  const parentFn = () => parent

  const { rerender } = render(<RepinnedExample parent={parentFn} forcePin />)

  expect(parent.addEventListener).toHaveBeenCalledTimes(1)

  parent.addEventListener.mockClear()

  rerender(<RepinnedExample parent={parentFn} />)

  expect(parent.addEventListener).toHaveBeenCalledTimes(1)
  expect(parent.addEventListener.mock.calls[0][0]).toBe('scroll')
})

test('Repinned toggling disabled on should remove event listeners', () => {
  const parent = getParent()
  const parentFn = () => parent

  const { rerender } = render(<RepinnedExample parent={parentFn} />)

  expect(parent.addEventListener.mock.calls[0][0]).toBe('scroll')
  expect(parent.addEventListener.mock.calls[1][0]).toBe('resize')

  rerender(<RepinnedExample parent={parentFn} disabled />)

  expect(parent.removeEventListener.mock.calls[0][0]).toBe('scroll')
  expect(parent.removeEventListener.mock.calls[1][0]).toBe('resize')
})

test('Repinned toggling disabled off should add event listeners', () => {
  const parent = getParent()
  const parentFn = () => parent

  const { rerender } = render(<RepinnedExample parent={parentFn} disabled />)

  expect(parent.addEventListener).not.toHaveBeenCalled()

  rerender(<RepinnedExample parent={parentFn} />)

  expect(parent.addEventListener.mock.calls[0][0]).toBe('scroll')
  expect(parent.addEventListener.mock.calls[1][0]).toBe('resize')
})

test('Repinned setHeightOffset updates the height in state', () => {
  const parent = getParent()
  let repinnedRef
  const setrepinnedRef = ref => {
    repinnedRef = ref
  }

  const { getByTestId } = render(
    <RepinnedExample parent={() => parent} repinnedRef={setrepinnedRef} />,
  )

  const height = '300px'
  repinnedRef.inner = {
    offsetHeight: height,
  }
  repinnedRef.setHeightOffset()
  expect(getByTestId('repinnedWrapper').style.height).toEqual(height)
})

test('Repinned handle scroll', async () => {
  let repinnedRef
  const setrepinnedRef = ref => {
    repinnedRef = ref
  }

  const onUnpinSpy = jest.fn()
  const onUnfixSpy = jest.fn()
  const onPinSpy = jest.fn()

  const parent = getParent()

  const { getByTestId } = render(
    <RepinnedExample
      onPin={onPinSpy}
      onUnpin={onUnpinSpy}
      onUnfix={onUnfixSpy}
      parent={() => parent}
      repinnedRef={setrepinnedRef}
    />,
  )

  parent.pageYOffset = 300
  repinnedRef.handleScroll()

  expect(getByTestId('RepinnedState')).toHaveTextContent('unpin')
  expect(onUnpinSpy).toHaveBeenCalledTimes(1)

  parent.pageYOffset = 180
  repinnedRef.handleScroll()

  expect(getByTestId('RepinnedState')).toHaveTextContent('pinned')
  expect(onPinSpy).toHaveBeenCalledTimes(1)

  parent.pageYOffset = 0
  repinnedRef.handleScroll()

  expect(getByTestId('RepinnedState')).toHaveTextContent('unfixed')
  expect(onUnfixSpy).toHaveBeenCalledTimes(1)
})

test('Repinned derives scrollY from scrollTop', async () => {
  let repinnedRef
  const setrepinnedRef = ref => {
    repinnedRef = ref
  }

  const parent = getParent()
  delete parent.pageYOffset
  parent.scrollTop = 30

  render(<RepinnedExample parent={() => parent} repinnedRef={setrepinnedRef} />)

  expect(repinnedRef.getScrollY()).toEqual(30)
})

test('Repinned getViewportHeight from window.innerHeight', async () => {
  let repinnedRef
  const setrepinnedRef = ref => {
    repinnedRef = ref
  }

  window.innerHeight = 1200

  render(<RepinnedExample repinnedRef={setrepinnedRef} />)

  expect(repinnedRef.getViewportHeight()).toEqual(1200)
})
