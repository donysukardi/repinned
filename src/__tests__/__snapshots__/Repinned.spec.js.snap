// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repinned component renders properly 1`] = `
<div
  style="min-height: 100vh;"
>
  <div
    data-testid="repinnedWrapper"
    style="height: 0px;"
  >
    <header
      data-testid="headerContainer"
    >
      <h1>
        This is header
      </h1>
      <div
        data-testid="RepinnedState"
      >
        unfixed
      </div>
    </header>
  </div>
  <div>
    This is some content
  </div>
</div>
`;
